### PR TITLE
[libc++][NFC] Rearrange functions in the synopsis comments of `<cmath>`

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
Currently, it's a bit hard to parse the `// C++17` comments nested in the `//  C99` group. This patch attempts to sort functions into 4 groups.
- C90: functions with names introduced in C89/90.
- C99: functions with names introduced C99, including `f` and `l` variants of C89/90 functions, except for 3D `hypot`.
- C++17: 3D `hypot` and C++17 math special functions.
- C++20: C++20 `lerp`.

The next group is expected to be C23, which will be added when C++ starts to merge C23 math functions.

More `arithmetic` and `floating_point` are used for simplification.

### CODE REVIEW

File: libcxx/include/cmath [hunk 1]
```diff
@@ -304,9 +314,17 @@ floating_point trunc (arithmetic x);
...
-constexpr float       lerp(float a, float b, float t) noexcept;                   // C++20
-constexpr double      lerp(double a, double b, double t) noexcept;                // C++20
-constexpr long double lerp(long double a, long double b, long double t) noexcept; // C++20
+//  C++17
+
+floating_point hypot(arithmetic x, arithmetic y, arithmetic z);
+
+floating_point hermite (unsigned n, arithmetic x);
+float          hermitef(unsigned n, float x);
+long double    hermitel(unsigned n, long double x);
```
Comment (frederick-vs-ja) [inline libcxx/include/cmath:323]: This will affect the convention of math special functions in synopsis comments. I think it will be better to show them consistently with most other functions.
@PaulXiCao Would you mind to follow this (probably pre-existing) convention?

**Comment by PaulXiCao** [comment]
The previous order was (almost?) an exact copy of the standard I believe (e.g. see https://eel.is/c++draft/c.math.syn). I see benefits to sticking to that order instead of defining a new one which might not be adhered to by new contributors.

Nonetheless, there are things to update, e.g. 
```cpp
//
// from main
// 
floating_point hypot (arithmetic x, arithmetic y);
float          hypotf(float x, float y);
long double    hypotl(long double x, long double y);

double       hypot(double x, double y, double z);                // C++17
float        hypot(float x, float y, float z);                   // C++17
long double  hypot(long double x, long double y, long double z); // C++17

//
// from the standard
//
constexpr floating-point-type hypot(floating-point-type x, floating-point-type y);
constexpr float hypotf(float x, float y);
constexpr long double hypotl(long double x, long double y);

// [[c.math.hypot3]](https://eel.is/c++draft/c.math.hypot3), three-dimensional hypotenuse
constexpr floating-point-type hypot(floating-point-type x, floating-point-type y,
                                    floating-point-type z);
```
I would also include `constexpr` as long as we support it.

I do not have an opinion yet what to do about the comments and links therein (e.g. `// [[c.math.hypot3]](https://eel.is/c++draft/c.math.hypot3), three-dimensional hypotenuse`). Not sure if there is really a benefit if we also include those.

**Comment by mordante** [comment]
> The previous order was (almost?) an exact copy of the standard I believe (e.g. see https://eel.is/c++draft/c.math.syn). I see benefits to sticking to that order instead of defining a new one which might not be adhered to by new contributors.

The goal is indeed to match the synopsis in the Standard to validate what we have and have not implemented.

> Nonetheless, there are things to update, e.g.
> 
> //
> // from main
> // 
> floating_point hypot (arithmetic x, arithmetic y);
> float          hypotf(float x, float y);
> long double    hypotl(long double x, long double y);
> 
> double       hypot(double x, double y, double z);                // C++17
> float        hypot(float x, float y, float z);                   // C++17
> long double  hypot(long double x, long double y, long double z); // C++17
> 
> //
> // from the standard
> //
> constexpr floating-point-type hypot(floating-point-type x, floating-point-type y);
> constexpr float hypotf(float x, float y);
> constexpr long double hypotl(long double x, long double y);
> 
> // [[c.math.hypot3]](https://eel.is/c++draft/c.math.hypot3), three-dimensional hypotenuse
> constexpr floating-point-type hypot(floating-point-type x, floating-point-type y,
>                                     floating-point-type z);

Since we have not implemented `constexpr hypot` (and friends) they should not be marked `constexpr` in the synopsis.
This is our way to determine what the status of our implementation is. Deviations from the Standard should mean we're missing features. This helps tracking the status of larger papers that are partly implemented.

Side note: This may not be the best way, but it's the best we got. We discussed other ideas in the past, but haven't found a better solution.

**Comment by ldionne** [comment]
Gentle ping on this -- do we want to move forward with some version of this, or close the PR?

**Comment by ldionne** [comment]
I'm going to tentatively close this since it's fairly old, please reopen if you still want to pursue this.

**Comment by frederick-vs-ja** [comment]
Sorry for the response. I won't pursue this PR, but I think the I'll make a larger cleanup for synopsis comments later.
